agent/dockerclient/dockerapi/docker_client.go

```diff
@@ -882,36 +883,34 @@ func getMetadataHealthCheck(dockerContainer *docker.Container) apicontainer.Heal
 
 // Listen to the docker event stream for container changes and pass them up
 func (dg *dockerGoClient) ContainerEvents(ctx context.Context) (<-chan DockerContainerChangeEvent, error) {
-	client, err := dg.dockerClient()
+	client, err := dg.sdkDockerClient()
 	if err != nil {
 		return nil, err
 	}
-	dockerEvents := make(chan *docker.APIEvents, dockerEventBufferSize)
-	events := make(chan *docker.APIEvents)
+	dockerEvents := make(<-chan events.Message, dockerEventBufferSize)
+	event := make(chan *events.Message)
 	buffer := NewInfiniteBuffer()
+	eventErr := make(<-chan error)
 
-	err = client.AddEventListener(dockerEvents)
-	if err != nil {
-		seelog.Errorf("DockerGoClient: unable to add a docker event listener: %v", err)
+	dockerEvents, eventErr = client.Events(ctx,  types.EventsOptions{})
+	select {
+	case err = <-eventErr:
+		seelog.Errorf("DockerGoClient: error while : %v", err)
 		return nil, err
```

Comment:
Correct me if I am wrong, but I think the `eventErr` is not used in this way according to the [SDK doc](https://github.com/moby/moby/blob/75fe41444009947d319f76e68b5e1f2124c79b4d/client/events.go#L19):

>Events returns a stream of events in the daemon. It's up to the caller to close the stream by cancelling the context. Once the stream has been completely read an io.EOF error will be sent over the error channel. If an error is sent all processing will be stopped. It's up to the caller to reopen the stream in the event of an error by reinvoking this method.

For your implementation, the err case will almost never be executed, because when the `eventErr` is returned, it's empty, and `select` will execute `default` case. But what if docker send back something via `eventErr` after you execute the `default` case? As SDK suggested, you probably need to:
1. Have a go routine listen to the `eventErr` channel.
2. Once we get error from `eventErr`, do something like logging and reopen the `dockerEvents` channel.

Reply:
Yes that makes sense. If, however, we get an error that is not EOF, does it make sense to reopen the channel and continue receiving events, or to propagate the error back up to return to ContainerEvents?